tools/buildsys/src/gomod.rs

```diff
@@ -0,0 +1,160 @@
+/*!
+Packages using the Go programming language may have upstream tar archives that
+include only the source code of the project, but not the source code of any
+dependencies.  The Go programming language promotes the use of "modules" for
+dependencies and projects adopting modules will provide go.mod and go.sum
+files.
+
+This module provides the ability to retrieve and validate the dependencies
+declared using Go modules given a tar archive containing a go.mod and go.sum.
+
+The location where dependencies are retrieved from are controlled by the
+standard environment variables employed by the Go tool: GOPROXY, GOSUMDB, and
+GOPRIVATE.
+
+ */
+
+pub(crate) mod error;
+use error::Result;
+
+use super::manifest;
+use duct::cmd;
+use snafu::{OptionExt, ResultExt};
+use std::env;
+use std::path::PathBuf;
+use std::process::Output;
+
+pub(crate) struct GoMod;
+
+impl GoMod {
+    pub(crate) fn vendor(
+        root_dir: &PathBuf,
+        package_dir: &PathBuf,
+        gomods: &[manifest::GoModule],
+    ) -> Result<Self> {
+        for g in gomods {
+            let input_path_arg = g.input.as_ref().context(error::InputFileSnafu)?;
+            let input_path = package_dir.join(input_path_arg);
+            if !input_path.is_file() {
+                return Err(error::Error::InputFileBad);
+            }
+            let mod_dir = g.mod_dir.as_ref().context(error::ModDirSnafu)?;
+            let output_dir_arg = g.output_dir.as_ref().context(error::OutputDirSnafu)?;
+            let output_dir = package_dir.join(output_dir_arg);
+            if output_dir.exists() && !output_dir.is_dir() {
+                return Err(error::Error::OutputDirBad);
+            }
+
+            // Our SDK and toolchain are picked by the external `cargo make` invocation.
+            let sdk = getenv("BUILDSYS_SDK_IMAGE")?;
+
+            // Several Go variables control proxying
+            let goproxy = go_env("GOPROXY").unwrap_or("".to_string());
+            let gosumdb = go_env("GOSUMDB").unwrap_or("".to_string());
+            let goprivate = go_env("GOPRIVATE").unwrap_or("".to_string());
+
+            let args = DockerGoArgs {
+                module_path: package_dir.clone(),
+                sdk_image: sdk,
+                go_mod_cache: root_dir.join(".gomodcache".to_string()),
+                command: format!(
+                    "mkdir -p {outdir}
+                    tar zxf {input} -C {outdir} &&
```

Comment:
You can just have `tar` figure out what kind of decompression to use:

```suggestion
                    tar xf {input} -C {outdir} &&
```